lib/endpoints/class-wp-json-controller.php

```diff
@@ -169,7 +169,120 @@ public function filter_response_by_context( $data, $context ) {
 	 * @return array
 	 */
 	public function get_item_schema() {
-		return array();
+		return $this->add_additional_fields_schema( array() );
+	}
+
+	/**
+	 * Add the values from additional fields to a data object
+	 * 
+	 * @param string $object_type
+	 * @param array  $object
+	 * @param WP_JSON_Request $request
+	 * @return array modified object with additional fields
+	 */
+	protected function add_additional_fields_to_object( $object, $request ) {
+
+		$additional_fields = $this->get_additional_fields();
+
+		foreach ( $additional_fields as $field_name => $field_options ) {
+
+			if ( ! $field_options['get_callback'] ) {
+				continue;
+			}
+
+			$object[$field_name] = call_user_func( $field_options['get_callback'], $object, $field_name, $request );
+		}
+
+		return $object;
+	}
+
+	protected function update_additional_fields_for_object( $object, $request ) {
+
+		$additional_fields = $this->get_additional_fields();
+
+		foreach ( $additional_fields as $field_name => $field_options ) {
+
+			if ( ! $field_options['update_callback'] ) {
+				continue;
+			}
+
+			// Don't run the update callbacks if the data wasn't passed in the request
+			if ( ! isset( $request[$field_name] ) ) {
+				continue;
+			}
+
+			$result = call_user_func( $field_options['update_callback'], $request[$field_name], $object, $field_name, $request );
```

Comment:
This seems to contradict your comment below:

>  I'm fine with us treating each field as atomic by itself though, rather than the whole set of fields as atomic (i.e. one failure doesn't break immediately).

I think the fields should be treat in insolation, if a given field fails to update, we shouldn't discriminate against other addition fields by other plugins etc.